### PR TITLE
[ENH] Add FreeSurfer aparc/aseg stats extractor

### LINKED ISSUE #392: [ENH] Migrate FreeSurfer extractor to CLI
Old scripts/utilities are [here](https://github.com/nipoppy/nipoppy/tree/fd553d7b29a686bb1e6b9eb611c25ffc52ec469c/nipoppy/extractors/freesurfer)

### PR DESCRIPTION
**Pull request by michellewang**
- Closes #392
- Related to #257 as this should only be run once for a single dataset

Changes proposed in this pull request:
- Add FreeSurfer extraction script
- Add descriptor and sample invocation files

Tested locally with Freesurfer 7.3.2 container and the following config:
```json
    {
            "NAME": "freesurfer_stats",
            "VERSION": "0.1.0",
            "CONTAINER_INFO": {
                "FILE": "[[NIPOPPY_DPATH_CONTAINERS]]/freesurfer_7.3.2.sif"
            },
            "CONTAINER_CONFIG": {
                "ENV_VARS": {
                    "FS_LICENSE": "<PATH>"
                },
                "ARGS": [
                    "--bind",
                    "[[NIPOPPY_DPATH_DERIVATIVES]]/freesurfer/7.3.2/output",
                    "--bind",
                    "[[NIPOPPY_DPATH_DERIVATIVES]]/freesurfer/7.3.2/idps/"
                ]
            },
            "STEPS": [
                {
                    "INVOCATION_FILE": "[[NIPOPPY_DPATH_INVOCATIONS]]/freesurfer_stats_and_qc-0.1.0.json",
                    "DESCRIPTOR_FILE": "[[NIPOPPY_DPATH_DESCRIPTORS]]/freesurfer_stats_and_qc-0.1.0.json",
                    "ANALYSIS_LEVEL": "group"
                }
            ]
        }
```

Output file headers (with default options):

<details>
<summary>IDP file</summary>

```
participant_id
session_id
Left-Lateral-Ventricle
Left-Inf-Lat-Vent
Left-Cerebellum-White-Matter
Left-Cerebellum-Cortex
Left-Thalamus
Left-Caudate
Left-Putamen
Left-Pallidum
3rd-Ventricle
4th-Ventricle
Brain-Stem
Left-Hippocampus
Left-Amygdala
CSF
Left-Accumbens-area
Left-VentralDC
Left-vessel
Left-choroid-plexus
Right-Lateral-Ventricle
Right-Inf-Lat-Vent
Right-Cerebellum-White-Matter
Right-Cerebellum-Cortex
Right-Thalamus
Right-Caudate
Right-Putamen
Right-Pallidum
Right-Hippocampus
Right-Amygdala
Right-Accumbens-area
Right-VentralDC
Right-vessel
Right-choroid-plexus
5th-Ventricle
WM-hypointensities
Left-WM-hypointensities
Right-WM-hypointensities
non-WM-hypointensities
Left-non-WM-hypointensities
Right-non-WM-hypointensities
Optic-Chiasm
CC_Posterior
CC_Mid_Posterior
CC_Central
CC_Mid_Anterior
CC_Anterior
BrainSegVol
BrainSegVolNotVent
lhCortexVol
rhCortexVol
CortexVol
lhCerebralWhiteMatterVol
rhCerebralWhiteMatterVol
CerebralWhiteMatterVol
SubCortGrayVol
TotalGrayVol
SupraTentorialVol
SupraTentorialVolNotVent
MaskVol
BrainSegVol-to-eTIV
MaskVol-to-eTIV
lhSurfaceHoles
rhSurfaceHoles
SurfaceHoles
EstimatedTotalIntraCranialVol
lh_bankssts_area
lh_caudalanteriorcingulate_area
lh_caudalmiddlefrontal_area
lh_cuneus_area
lh_entorhinal_area
lh_fusiform_area
lh_inferiorparietal_area
lh_inferiortemporal_area
lh_isthmuscingulate_area
lh_lateraloccipital_area
lh_lateralorbitofrontal_area
lh_lingual_area
lh_medialorbitofrontal_area
lh_middletemporal_area
lh_parahippocampal_area
lh_paracentral_area
lh_parsopercularis_area
lh_parsorbitalis_area
lh_parstriangularis_area
lh_pericalcarine_area
lh_postcentral_area
lh_posteriorcingulate_area
lh_precentral_area
lh_precuneus_area
lh_rostralanteriorcingulate_area
lh_rostralmiddlefrontal_area
lh_superiorfrontal_area
lh_superiorparietal_area
lh_superiortemporal_area
lh_supramarginal_area
lh_frontalpole_area
lh_temporalpole_area
lh_transversetemporal_area
lh_insula_area
lh_WhiteSurfArea_area
rh_bankssts_area
rh_caudalanteriorcingulate_area
rh_caudalmiddlefrontal_area
rh_cuneus_area
rh_entorhinal_area
rh_fusiform_area
rh_inferiorparietal_area
rh_inferiortemporal_area
rh_isthmuscingulate_area
rh_lateraloccipital_area
rh_lateralorbitofrontal_area
rh_lingual_area
rh_medialorbitofrontal_area
rh_middletemporal_area
rh_parahippocampal_area
rh_paracentral_area
rh_parsopercularis_area
rh_parsorbitalis_area
rh_parstriangularis_area
rh_pericalcarine_area
rh_postcentral_area
rh_posteriorcingulate_area
rh_precentral_area
rh_precuneus_area
rh_rostralanteriorcingulate_area
rh_rostralmiddlefrontal_area
rh_superiorfrontal_area
rh_superiorparietal_area
rh_superiortemporal_area
rh_supramarginal_area
rh_frontalpole_area
rh_temporalpole_area
rh_transversetemporal_area
rh_insula_area
rh_WhiteSurfArea_area
```

</details>

<details>
<summary>QC file</summary>

```
participant_id
session_id
n_holes-lh_white
n_holes-rh_white
n_holes-lh_pial
n_holes-rh_pial
gray_white_cnr-norm_mgz-lh
gray_white_cnr-norm_mgz-rh
gray_csf_cnr-norm_mgz-lh
gray_csf_cnr-norm_mgz-rh
```
</details>

## Checklist (for reviewers)
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (e.g. `[BUG]`, `[DOC]`, `[ENH]`, `[MAINT]`)\
Refer to [NumPy Development Guide](https://numpy.org/doc/stable/dev/development_workflow.html#writing-the-commit-message) for a full list
- [x] PR links to GitHub issue with mention `Closes #XXXX`
- [x] Tests pass
- [x] Checks pass

For new features:
- [ ] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions

### CODE REVIEW

File: nipoppy/data/examples/sample_invocations/freesurfer_stats_and_qc-0.1.0.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+    "script_path": "[[NIPOPPY_DPATH_PROC]]/extractors/freesurfer_stats_and_qc.py",
```
Comment (michellewang) [inline nipoppy/data/examples/sample_invocations/freesurfer_stats_and_qc-0.1.0.json:2]: Not sure where we want to put the actual script in the Nipoppy dataset...

File: nipoppy/workflows/extractor.py [hunk 2]
```diff
@@ -123,8 +139,8 @@ def run_single(self, participant_id: str, session_id: str):
             participant_id=participant_id,
             session_id=session_id,
             bind_paths=[
-                self.layout.dpath_sourcedata,
-                self.layout.dpath_bids,
+                self.dpath_pipeline_idp,
+                self.dpath_pipeline_output,
```
Comment (michellewang) [inline nipoppy/workflows/extractor.py:143]: Fixing wrong bind paths for the container

File: nipoppy/workflows/extractor.py [hunk 3]
```diff
@@ -91,6 +91,22 @@ def dpath_pipeline(self) -> Path:
...
+        )
+
+    @cached_property
+    def dpath_pipeline_idp(self) -> Path:
+        """Return the path to the pipeline's IDP directory."""
+        return self.layout.get_dpath_pipeline_idp(
+            pipeline_name=self.proc_pipeline_info.NAME,
+            pipeline_version=self.proc_pipeline_info.VERSION,
+        )
+
```
Comment (michellewang) [inline nipoppy/workflows/extractor.py:109]: These do not depend on `dpath_pipeline` so they each have to be overriden to use `proc_pipeline_info`, which is annoying so I made an issue to refactor this

File: nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/invocation.json [hunk 4]
```diff
@@ -1,11 +1,11 @@
 {
     "script_path": "[[NIPOPPY_DPATH_PIPELINES]]/freesurfer_stats_and_qc-0.1.0/script.py",
     "input_dir_path": "[[NIPOPPY_DPATH_PIPELINE_OUTPUT]]",
-    "output_stats_file_path": "[[NIPOPPY_DPATH_PIPELINE_IDP]]/freesurfer_stats.tsv",
+    "output_stats_file_path": "[[NIPOPPY_DPATH_PIPELINE_IDP]]/freesurfer_stats-thickness.tsv",
```
Comment (nikhil153) [inline nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/invocation.json:4]: Should we add parcelation tag as well? 
e.g. `DKT`?

File: nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/script.py [hunk 5]
```diff
@@ -379,10 +379,30 @@ def run_single_stats(
...
+                        first = df_to_concat[col]
+                    else:
+                        common_participants = list(
+                            set(first.index).intersection(set(df_to_concat.index))
+                        )
+                        if first.loc[common_participants].equals(
+                            df_to_concat.loc[common_participants, col]
+                        ):
+                            df_to_concat.drop(col, axis="columns", inplace=True)
+                            print(f"Dropped duplicate column {col} from aparc file.")
```
Comment (nikhil153) [inline nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/script.py:397]: Why would there be duplicates?

**Review by nikhil153 — APPROVED** [review]
Not 100% sure how to specify `parcelations`. For now may be we stick to filename itself.

File: nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/script.py (hunk 5 above)
Comment (michellewang) [reply nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/script.py:397]: When I ran this on FreeSurfer 6 I got duplicate columns (with identical contents) in the left/right aparc files. Like `eTIV`, `BrainSegVolNotVentSurf`, and `SupraTentorialVolNotVentVox`

File: nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/invocation.json [hunk 6]
```diff
@@ -0,0 +1,11 @@
+{
+    "script_path": "[[NIPOPPY_DPATH_PIPELINES]]/freesurfer_stats_and_qc-0.1.0/script.py",
+    "input_dir_path": "[[NIPOPPY_DPATH_PIPELINE_OUTPUT]]",
+    "output_stats_file_path": "[[NIPOPPY_DPATH_PIPELINE_IDP]]/freesurfer_stats-thickness.tsv",
+    "output_qc_file_path": "[[NIPOPPY_DPATH_PIPELINE_IDP]]/freesurfer_qc.tsv",
+    "mode": "multi",
+    "container": "[[NIPOPPY_CONTAINER_COMMAND]] [[NIPOPPY_FPATH_CONTAINER]]",
+    "aseg_args": " --skip",
+    "aparc_args": " --skip --meas=thickness",
```
Comment (nikhil153) [inline nipoppy/data/examples/sample_pipelines/freesurfer_stats_and_qc-0.1.0/invocation.json:9]: i think `meas` needs to be a tag as well. `thickness` and `surface-area` are both common and would be easier to parse as a key-val

**Pull request merged**